tags/ax25tools/1.0.3/man/netromd.8
.TH NETROMD 8 "20 August 1996" Linux "Linux System Managers Manual"
.SH NAME
netromd \- Send and receive NET/ROM routing messages
.SH SYNOPSIS
.B netromd [-c] [-d] [-i] [-l] [-p pause] [-q quality] [-t interval] [-v]
.SH DESCRIPTION
.LP
For a NET/ROM based network to operate correctly, a periodic broadcast of
routing information needs to occur. Typically this occurs once every hour on
every port which is expected to carry NET/ROM traffic. The purpose of
.B netromd
is to send and receive NET/ROM routing broadcasts. To operate correctly a
set of parameters that corresponds to each AX.25 port needs to be passed to
the program. This information is encoded in a configuration file, by default
which is /etc/ax25/nrbroadcast with each line representing one
port, see the manual page for
.BR nrbroadcast (5).
.LP
To cut down the length of these routing broadcasts, only the information
about the highest quality neighbour for a particular node is transmitted.
The transmission is also limited to those node that have a certain minimum
value in their obsolesence count, this value is decremented every time a
routing broadcast is transmitted, and is refreshed by receiving a routing
broadcast which contains that particular node.
.LP
The value of the default quality is traditionally assigned a value that
represents the quality of the radio links on that port. A higher number
representing better radio links with 255 (the maximum) reserved for wire
connections. The practise in the UK is to set the default quality to a low
value, typically 10, and manually set up the trusted neighbouring nodes in
the neighbour list manually. The worst quality for auto-updates value is a
way to filter out low quality (ie distant) nodes.
.LP
The verbose flag may be either 0 or 1, representing no and yes. By
specifying no, the program will only generate a routing message containing
information about the node on which it is running, by specifying the yes
option, all the information in the nodes routing tables will be transmitted.
The quality advertised for the other node callsigns on this machine may be
set using the \-q option.
.LP
Between each transmission
.B netromd
pauses for five seconds (default) in order to avoid flooding the channels
that it must broadcast on. The value of this delay is settable with the \-p
option.
.SH OPTIONS
.TP 16
.BI \-c
Forces strict compliance to Software 2000 specifications. At present this
only determines how node mnemonics with lower case characters will be handled.
With compliance enabled mixed case node mnemonics will be ignored. The default
is to accept node mnemonics of mixed case.
.TP 16
.BI \-d
Switches on debugging messages, the default is off. Logging must be enabled
for them to be output.
.TP 16
.BI \-i
Transmit a routing broadcast immediately, the default is to wait for the
time interval to elapse before transmitting the first routing broadcast.
.TP 16
.BI \-l
Enables logging of errors and debug messages to the system log. The default
is off.
.TP 16
.BI "\-p pause"
Sets the delay between transmissions of individual routing broadcast
packets. The default is five seconds.
.TP 16
.BI "\-q quality"
Sets the quality of the subsidiary nodes relative to the main node. The
default is 255.
.TP 16
.BI "\-t interval"
The time interval between routing broadcasts, in minutes. The default is 60
minutes.
.TP 16
.BI \-v
Display the version.
.SH FILES
.nf
/proc/net/nr_neigh
.br
/proc/net/nr_nodes
.br
/etc/ax25/axports
.br
/etc/ax25/nrbroadcast
.fi
.SH "SEE ALSO"
.BR ax25 (4),
.BR axports (5),
.BR nrbroadcast (5),
.BR netrom (4),
.BR nrparms (8).
.SH AUTHOR
Jonathan Naylor G4KLX <g4klx@g4klx.demon.co.uk>


ax25apps/NEWS
NEWS for ax25-apps
==================

These programs are still being tested, you should expect some quirks
especially with compiling for a little while longer




tags/ax25tools/1.0.2/netrom/netromd.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <syslog.h>
#include <signal.h>
#include <time.h>
#include <sys/time.h>

#include <config.h>

#include <sys/types.h>
#include <sys/ioctl.h>

#include <sys/socket.h>
#include <netinet/in.h>

#include <net/if.h>

#ifdef __GLIBC__ 
#include <net/ethernet.h>
#else
#include <linux/if_ether.h>
#endif

#include <netax25/ax25.h>
#include <netrom/netrom.h>
#include <netrose/rose.h>

#include <netax25/axlib.h>
#include <netax25/axconfig.h>
#include <netax25/daemon.h>
#include <netax25/nrconfig.h>

#include "../pathnames.h"
#include "netromd.h"

struct port_struct port_list[20];

int port_count = FALSE;
int compliant  = FALSE;
int logging    = FALSE;
int debug      = 0;

ax25_address my_call;
ax25_address node_call;

static void terminate(int sig)
{
	if (logging) {
		syslog(LOG_INFO, "terminating on SIGTERM\n");
		closelog();
	}
	
	exit(0);
}

static int bcast_config_load_ports(void)
{
	char buffer[255], port[32], *s;
	FILE *fp;
	
	if ((fp = fopen(CONF_NETROMD_FILE, "r")) == NULL) {
		fprintf(stderr, "netromd: cannot open config file\n");
		return -1;
	}
		
	while (fgets(buffer, 255, fp) != NULL) {
		if ((s = strchr(buffer, '\n')) != NULL)
			*s = '\0';

		if (strlen(buffer) == 0 || buffer[0] == '#')
			continue;

		if (sscanf(buffer, "%s %d %d %d %d",
				port,
				&port_list[port_count].minimum_obs,
				&port_list[port_count].default_qual,
				&port_list[port_count].worst_qual,
				&port_list[port_count].verbose) == -1) {
			fprintf(stderr, "netromd: unable to parse: %s", buffer);
			return -1;
		}

		if (ax25_config_get_addr(port) == NULL) {
			fprintf(stderr, "netromd: invalid port name - %s\n", port);
			return -1;
		}

		port_list[port_count].port   = strdup(port);
		port_list[port_count].device = strdup(ax25_config_get_dev(port_list[port_count].port));

		if (port_list[port_count].minimum_obs < 0 || port_list[port_count].minimum_obs > 6) {
			fprintf(stderr, "netromd: invalid minimum obsolescence\n");
			return -1;
		}

		if (port_list[port_count].default_qual < 0 || port_list[port_count].default_qual > 255) {
			fprintf(stderr, "netromd: invalid default quality\n");
			return -1;
		}

		if (port_list[port_count].worst_qual < 0 || port_list[port_count].worst_qual > 255) {
			fprintf(stderr, "netromd: invalid worst quality\n");
			return -1;
		}

		if (port_list[port_count].verbose != 0 && port_list[port_count].verbose != 1) {
			fprintf(stderr, "netromd: invalid verbose setting\n");
			return -1;
		}
		
		port_count++;
	}
	
	fclose(fp);

	if (port_count == 0)
		return -1;
	
	return 0;	
}

int main(int argc, char **argv)
{
	unsigned char buffer[512];
	int size, s, i;
	struct sockaddr sa;
	socklen_t asize;
	struct timeval timeout;
	time_t timenow, timelast;
	int interval = 3600;
	int localval = 255;
	int pause    = 5;
	fd_set fdset;

	time(&timelast);

	while ((i = getopt(argc, argv, "cdilp:q:t:v")) != -1) {
		switch (i) {
			case 'c':
				compliant = TRUE;
				break;
			case 'd':
				debug++;
				break;
			case 'i':
				timelast = 0;
				break;
			case 'l':
				logging = TRUE;
				break;
			case 'p':
				pause = atoi(optarg);
				if (pause < 1 || pause > 120) {
					fprintf(stderr, "netromd: invalid pause value\n");
					return 1;
				}
				break;
			case 'q':
				localval = atoi(optarg);
				if (localval < 10 || localval > 255) {
					fprintf(stderr, "netromd: invalid local quality\n");
					return 1;
				}
				break;
			case 't':
				interval = atoi(optarg) * 60;
				if (interval < 60 || interval > 7200) {
					fprintf(stderr, "netromd: invalid time interval\n");
					return 1;
				}
				break;
			case 'v':
				printf("netromd: %s\n", VERSION);
				return 0;
			case '?':
			case ':':
				fprintf(stderr, "usage: netromd [-d] [-i] [-l] [-q quality] [-t interval] [-v]\n");
				return 1;
		}
	}

	signal(SIGTERM, terminate);

	if (ax25_config_load_ports() == 0) {
		fprintf(stderr, "netromd: no AX.25 ports defined\n");
		return 1;
	}
	
	if (nr_config_load_ports() == 0) {
		fprintf(stderr, "netromd: no NET/ROM ports defined\n");
		return 1;
	}

	if (bcast_config_load_ports() == -1) {
		fprintf(stderr, "netromd: no NET/ROM broadcast ports defined\n");
		return 1;
	}

	ax25_aton_entry(nr_config_get_addr(NULL), (char *)&my_call);
	ax25_aton_entry("NODES", (char *)&node_call);
	
	if ((s = socket(PF_PACKET, SOCK_PACKET, htons(ETH_P_AX25))) == -1) {
		perror("netromd: socket");
		return 1;
	}

	if (!daemon_start(TRUE)) {
		fprintf(stderr, "netromd: cannot become a daemon\n");
		return 1;
	}
	
	if (logging) {
		openlog("netromd", LOG_PID, LOG_DAEMON);
		syslog(LOG_INFO, "starting");
	}

	for (;;) {
		FD_ZERO(&fdset);
		FD_SET(s, &fdset);

		timeout.tv_sec  = 30;
		timeout.tv_usec = 0;
	
		if (select(s + 1, &fdset, NULL, NULL, &timeout) == -1)
			continue;		/* Signal received ? */

		if (FD_ISSET(s, &fdset)) {
			asize = sizeof(sa);

			if ((size = recvfrom(s, buffer, sizeof(buffer), 0, &sa, &asize)) == -1) {
				if (logging) {
					syslog(LOG_ERR, "recvfrom: %m");
					closelog();
				}
				return 1;
			}

			if (ax25_cmp((ax25_address *)(buffer + 1), &node_call) == 0 &&
			    ax25_cmp((ax25_address *)(buffer + 8), &my_call)   != 0 && 
			    buffer[16] == NETROM_PID && buffer[17] == NODES_SIG) {
				for (i = 0; i < port_count; i++) {
					if (strcmp(port_list[i].device, sa.sa_data) == 0) {
						if (debug && logging)
							syslog(LOG_DEBUG, "receiving NODES broadcast on port %s from %s\n", port_list[i].port, ax25_ntoa((ax25_address *)(buffer + 8)));
						receive_nodes(buffer + 18, size - 18, (ax25_address *)(buffer + 8), i);
						break;
					}
				}
			}
		}

		time(&timenow);

		if ((timenow - timelast) > interval) {
			timelast = timenow;
			transmit_nodes(localval, pause);
		}
	}
}
